### PR TITLE
LBANN Tests

### PR DESCRIPTION
**Pull request by forsyth2**
Initial tests for LBANN performance and accuracy.

### CODE REVIEW

**Comment by forsyth2** [comment]
@DylanMcKinney Here's the tests.

File: .gitignore [hunk 1]
```diff
@@ -7,3 +7,4 @@
 
 # Can also ignore all directories and files in a directory.
 # tmp/**/* 
+build
```
Comment (forsyth2) [inline .gitignore]: Add a new line.

File: scripts/test.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  return '--optimizer=../model_zoo/optimizers/%s.prototext' % optimizer_name
+
+def executable(cluster):
+  return '../build/%s/model_zoo/lbann' % cluster
+
+def run_lbann(model_name, reader_name, optimizer_name, output_file_name):
+  if (model_name == None or reader_name == None or optimizer_name == None or output_file_name == None):
+    raise Exception("Invalid number of arguments")
+  else:
+    os.system('hostname > %s' % output_file_name)
```
Comment (forsyth2) [inline scripts/test.py]: Inefficient to write to file just to get this information? Would it be better to use a subprocess?

File: scripts/test.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    e = executable(cluster)
+    m = model(model_name)
+    r = reader(reader_name)
+    o = optimizer(optimizer_name)
+    # N => number of nodes                                                         
+    # p => partition
+    # t => timeout period, in minutes
+    # n => number of processes to run. MPI Rank. How many models being made.
+    # num-epochs => The number of epochs
+    os.system('salloc -N 1 -p pbatch -t 30 srun -n 2 %s %s %s %s --num_epochs=5 > %s' % (e, m, r, o, output_file_name))
```
Comment (forsyth2) [inline scripts/test.py]: Remove `-t 30` for longer running times? Are there any options besides model, reader, optimizer we want to specify?

File: scripts/test.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      performance_dict[model]['epoch_times'].append(float(m.group(2)))
+      performance_dict[model]['mean_minibatch_times'].append(float(m.group(4)))
+      performance_dict[model]['mins'].append(float(m.group(6)))
+      performance_dict[model]['maxs'].append(float(m.group(8)))
+      performance_dict[model]['stdevs'].append(float(m.group(10)))
+    a = re.match('(Model [0-9]+) @[0-9]+ testing steps external validation categorical accuracy: ([0-9]+(\.[0-9]+)*)', line)
+    if a:
+      model = a.group(1)
+      if model not in performance_dict.keys():
+        performance_dict[model] = {
```
Comment (forsyth2) [inline scripts/test.py]: Make an `accuracy_dict` and put accuracy in that one?

File: scripts/test.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  run_lbann(model_name, reader_name, optimizer_name, output_file_name)
+  return get_performance(output_file_name)
+
+class LBANNTest(unittest.TestCase):
+  def test_performance(self):
+    performance = main(model_name='model_mnist_multi', reader_name='data_reader_mnist', optimizer_name='opt_adagrad', output_file_name='output.txt')
+    # performance = get_performance(output_file_name='output.txt')
+    for model in performance.keys():
+      for epoch_time in performance[model]['epoch_times']:
+        self.assertTrue(epoch_time < 67)
```
Comment (forsyth2) [inline scripts/test.py]: What values should we be testing against?

File: scripts/test.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      performance_dict[model]['accuracies'].append(float(a.group(2)))
+  output_file.close()
+  return performance_dict
+
+def main(model_name, reader_name, optimizer_name, output_file_name):
+  run_lbann(model_name, reader_name, optimizer_name, output_file_name)
+  return get_performance(output_file_name)
+
+class LBANNTest(unittest.TestCase):
+  def test_performance(self):
```
Comment (forsyth2) [inline scripts/test.py]: Maybe we should keep performance and accuracy together per test. However, each test will test a different model.
Perhaps rename this `test_mnist_performance` and add `test_resnet_performance` and `test_alexnet_performance`. Abstract out common code.

File: scripts/test.py (hunk 3 above)
Comment (bvanessen) [reply scripts/test.py]: Eventually we are going to want to vary the number of nodes and processes, which also dictates the processes per node.

**Comment by davidHysom** [comment]
You may have noticed, I've instituted a convention that prototext file names are of the form:

  model_XX.prototext

  data_reader_XX.prototext

  model_XX.prototext

If we want to adopt this convention, I would suggest that,

when specifying files for the tests, users should not have

to specify the 'model_', etc, portion of the names.

OTOH, perhaps my convention is overly restrictive.


________________________________

**Comment by forsyth2** [comment]
@DylanMcKinney I added some minor improvements to the test script but I still need to run the tests to make sure nothing broke.

File: .gitignore (hunk 1 above)
Comment (forsyth2) [reply .gitignore]: Added.
